Feature prune join output

sopel39 opened the pull request:
Recreating after closing (removed branch) https://github.com/prestodb/presto/pull/6465 by mistake

FYI: @martint @dain

Partial micro benchmark results:
```
    Partial results:
    Benchmark                                             (buildHashEnabled)  (hashColumns)  (matchRate)  (outputColumns)  Mode  Cnt    Score   Error  Units
    BenchmarkHashBuildAndJoinOperators.benchmarkJoinHash                true         bigint            2          varchar  avgt   60  157,080 ± 3,820  ms/op
    BenchmarkHashBuildAndJoinOperators.benchmarkJoinHash                true         bigint            2           bigint  avgt   60   90,912 ± 1,954  ms/op
    BenchmarkHashBuildAndJoinOperators.benchmarkJoinHash                true         bigint            2              all  avgt   60  203,645 ± 3,665  ms/op
```
`all` is a baseline. When you require just some columns the join time reduces significantly.

losipiuk commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java:
```diff
@@ -162,10 +162,7 @@ public PlanNode getRight()
-                .addAll(left.getOutputSymbols())
-                .addAll(right.getOutputSymbols())
-                .build();
+        return outputSymbols;
```
this should be part of commit which introduced the `outputSymbols` field.

sopel39 replied:
neh, now you can pull first commit without the latter. If you move it to first commit, tests will fail

sopel39 commented on presto-main/src/test/java/com/facebook/presto/operator/BenchmarkHashBuildAndJoinOperators.java:
```diff
@@ -257,8 +257,8 @@ public LookupSourceFactory benchmarkBuildHash(BuildContext buildContext)
 
-        if (!operator.isFinished()) {
-            throw new AssertionError("Expected hash build operator to be finished");
+        if (!hashBuilderOperatorFactory.getLookupSourceFactory().createLookupSource().isDone()) {
```
Benchmark was broken before. `LookupSource` is cached in `hashBuilderOperatorFactory`, so it's not recreated. Anyway, we need to benchmark hash building and part of it is `LookupSource` creation

sopel39 commented on presto-benchmark/src/main/java/com/facebook/presto/benchmark/HashBuildAndJoinBenchmark.java:
```diff
@@ -92,7 +92,7 @@ from lineitem join orders using (orderkey)
             hashChannel = Optional.of(2);
```
Yeah but it is in `JoinProbeCompiler`

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/gen/JoinProbeCompiler.java:
```diff
@@ -94,15 +94,20 @@ public OperatorFactory compileJoinOperatorFactory(int operatorId,
             JoinType joinType)
     {
         try {
+            List<Integer> probeOutputChannelIndexes = probeOutputChannels.orElse(IntStream.range(0, probeTypes.size())
```
done

sopel39 commented on presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java:
```diff
@@ -307,21 +308,26 @@ private PagesIndexOrdering createPagesIndexComparator(List<Integer> sortChannels
+        return createPagesHashStrategy(joinChannels, hashChannel, Optional.empty());
+    }
+
+    public PagesHashStrategy createPagesHashStrategy(List<Integer> joinChannels, Optional<Integer> hashChannel, Optional<List<Integer>> outputChannels)
```
I prefer to leave it optional here for now than modifying all classes that use `PagesIndex` and `JoinCompiler`. It will be easier to rebase and resolve conflicts while exposing the parameter.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java:
```diff
@@ -307,21 +308,26 @@ private PagesIndexOrdering createPagesIndexComparator(List<Integer> sortChannels
+        return new SimplePagesHashStrategy(types, outputChannels.orElse(rangeList(types.size())), ImmutableList.<List<Block>>copyOf(channels), joinChannels, hashChannel);
     }
 
     public Supplier<LookupSource> createLookupSourceSupplier(
```
not everybody uses `outputChannels` parameters (e.g: `IndexSnapshotBuilder`)

sopel39 commented on presto-main/src/main/java/com/facebook/presto/operator/SimplePagesHashStrategy.java:
```diff
@@ -68,12 +70,12 @@ public long getSizeInBytes()
             Block block = channel.get(blockIndex);
-            type.appendTo(block, position, pageBuilder.getBlockBuilder(outputChannelOffset));
-            outputChannelOffset++;
+            type.appendTo(block, position, pageBuilder.getBlockBuilder(outputChannelOffset + (pageBuilderOutputChannel++)));
```
done

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/EffectivePredicateExtractor.java:
```diff
@@ -224,9 +224,11 @@ public Expression visitJoin(JoinNode node, Void context)
-            joinConjuncts.add(new ComparisonExpression(ComparisonExpression.Type.EQUAL,
-                    clause.getLeft().toSymbolReference(),
-                    clause.getRight().toSymbolReference()));
+            if (node.getOutputSymbols().contains(clause.getLeft()) && node.getOutputSymbols().contains(clause.getRight())) {
```
nice catch

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java:
```diff
@@ -319,16 +309,41 @@ public PlanWithProperties visitJoin(JoinNode node, HashComputationSet parentPref
+                Optional<Symbol> rightHashSymbol)
+        {
+            // retain only hash symbols preferred by parent nodes
+            allHashSymbols.keySet().retainAll(parentPreference.getHashes());
```
done

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java:
```diff
@@ -319,16 +309,41 @@ public PlanWithProperties visitJoin(JoinNode node, HashComputationSet parentPref
+            // retain only hash symbols preferred by parent nodes
+            allHashSymbols.keySet().retainAll(parentPreference.getHashes());
+
+            ImmutableList<Symbol> outputSymbols = ImmutableList.<Symbol>builder()
```
done

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java:
```diff
@@ -198,7 +199,23 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Set<Symbol>> context)
-            return new JoinNode(node.getId(), node.getType(), left, right, node.getCriteria(), node.getFilter(), node.getLeftHashSymbol(), node.getRightHashSymbol());
+            List<Symbol> outputSymbols;
+            if (node.getCriteria().isEmpty() && !node.getFilter().isPresent() && node.getType() == INNER) {
+                // do not prune nested joins output since it is not supported
```
done. This is more implementation specific (e.g: nested loop doesn't support filter), but I agree one condition method is better.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java:
```diff
@@ -198,7 +199,23 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Set<Symbol>> context)
+                Set<Symbol> seenSymbol = new HashSet<>();
+                outputSymbols = node.getOutputSymbols().stream()
+                        .filter(symbol -> context.get().contains(symbol))
+                        .filter(seenSymbol::add)
```
sure they can

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java:
```diff
@@ -345,17 +353,35 @@ public Void visitJoin(JoinNode node, Set<Symbol> boundSymbols)
+                    .map(symbol -> node.getOutputSymbols().indexOf(symbol))
+                    .collect(toImmutableList());
+
+            boolean allLeftBeforeRight = node.getLeft().getOutputSymbols().stream()
```
performance irrelevant

sopel39 commented:
addressed comments, https://github.com/prestodb/presto/tree/9dc6a9e42aa78f8a2663a20ece95c0f72fe07734 and https://github.com/prestodb/presto/tree/15a8c8648ca497961fb420c7056b7eec5ba7a5d0 are part of other PR (skip review)

sopel39 commented:
also added plan test for pruning

sopel39 commented:
@Lewuathe remove changes requested flag if OK

sopel39 commented:
rebased on newest master. FYI @martint 

Pruning join output columns gives huge performance improvements (50% is common, see benchmarks in the first comment) for joins where only some columns are used later in the query. This is very typical case.

Case 1:
```
query:
SELECT a.x FROM a, b WHERE a.y = b.y AND a.z = b.z AND a.m > 10
Symbols produced by LookupJoinNode in current code:
a.x, a.y, b.y, a.z, b.z, a.m, hash(a.y), hash(b.y), hash(a.z), hash(b.z)
Symbols produced by LookupJoinNode after this PR:
a.x
``` 

Simpler Case 2:
```
query:
SELECT a.x FROM a, b WHERE a.x = b.x
Symbols produced by LookupJoinNode in current code:
a.x, b.x, hash(a.x), hash(b.x)
Symbols produced by LookupJoinNode after this PR:
a.x
```

As you can see, memory copying is greatly reduced. This overhead is especially important for joins that quickly evaluate matching rows (e.g: BIGINT equi-clause)

sopel39 commented:
Actual query:
```
SELECT count(*) FROM lineitem l, part p, supplier s, partsupp ps, orders o, nation n WHERE s.suppkey = l.suppkey AND ps.suppkey = l.suppkey AND ps.partkey = l.partkey AND p.partkey = l.partkey AND o.orderkey = l.orderkey AND s.nationkey = n.nationkey;
```
Before:
```
presto:default> SELECT count(*) FROM lineitem l, part p, supplier s, partsupp ps, orders o, nation n WHERE s.suppkey = l.suppkey AND ps.suppkey = l.suppkey AND ps.partkey = l.partkey AND p.partkey = l.partkey AND o.orderkey = l.orderkey AND s.nationkey = n.nationkey;
  _col0
---------
 6001215
(1 row)

Query 20161219_130526_00007_ctqww, FINISHED, 1 node
Splits: 164 total, 164 done (100,00%)
0:18 [8,51M rows, 215MB] [462K rows/s, 11,7MB/s]

- InnerJoin[("nationkey" = "nationkey_14")][$hashvalue, $hashvalue_48] => [nationkey:bigint, $hashvalue:bigint, nationkey_14:bigint, $hashvalue_48:bigint]
    - LocalExchange[ROUND_ROBIN] () => nationkey:bigint, $hashvalue:bigint
        - RemoteExchange[REPARTITION][$hashvalue_22] => nationkey:bigint, $hashvalue_22:bigint
            - Project => [nationkey:bigint, $hashvalue_47:bigint]
                    $hashvalue_47 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("nationkey"), 0))
                - InnerJoin[("orderkey" = "orderkey_11")][$hashvalue_23, $hashvalue_44] => [nationkey:bigint, orderkey:bigint, $hashvalue_23:bigint, orderkey_11:bigint, $hashvalue_44:bigint]
                    - LocalExchange[ROUND_ROBIN] () => nationkey:bigint, orderkey:bigint, $hashvalue_23:bigint
                        - RemoteExchange[REPARTITION][$hashvalue_24] => nationkey:bigint, orderkey:bigint, $hashvalue_24:bigint
                            - Project => [nationkey:bigint, orderkey:bigint, $hashvalue_43:bigint]
                                    $hashvalue_43 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("orderkey"), 0))
                                - InnerJoin[("partkey" = "partkey_7") AND ("suppkey" = "suppkey_8")][$hashvalue_37, $hashvalue_38] => [partkey:bigint, nationkey:bigint, suppkey:bigint, orderkey:bigint, $hashvalue_37:bigint, suppkey_8:bigint, partkey_7:bigint, $hashvalue_38:bigint]
                                    - Project => [partkey:bigint, nationkey:bigint, suppkey:bigint, orderkey:bigint, $hashvalue_37:bigint]
                                            $hashvalue_37 := "combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("partkey"), 0)), COALESCE("$operator$hash_code"("suppkey"), 0))
                                        - InnerJoin[("suppkey" = "suppkey_3")][$hashvalue_25, $hashvalue_34] => [partkey:bigint, orderkey:bigint, suppkey:bigint, $hashvalue_25:bigint, nationkey:bigint, suppkey_3:bigint, $hashvalue_34:bigint]
                                            - LocalExchange[ROUND_ROBIN] () => partkey:bigint, orderkey:bigint, suppkey:bigint, $hashvalue_25:bigint
                                                - RemoteExchange[REPARTITION][$hashvalue_26] => partkey:bigint, orderkey:bigint, suppkey:bigint, $hashvalue_26:bigint
                                                    - Project => [partkey:bigint, orderkey:bigint, suppkey:bigint, $hashvalue_33:bigint]
                                                            $hashvalue_33 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("suppkey"), 0))
                                                        - InnerJoin[("partkey" = "partkey_0")][$hashvalue_27, $hashvalue_30] => [partkey:bigint, orderkey:bigint, suppkey:bigint, $hashvalue_27:bigint, partkey_0:bigint, $hashvalue_30:bigint]

```

After:
```
presto:default> SELECT count(*) FROM lineitem l, part p, supplier s, partsupp ps, orders o, nation n WHERE s.suppkey = l.suppkey AND ps.suppkey = l.suppkey AND ps.partkey = l.partkey AND p.partkey = l.partkey AND o.orderkey = l.orderkey AND s.nationkey = n.nationkey;
  _col0
---------
 6001215
(1 row)

Query 20161219_130105_00006_26ksu, FINISHED, 1 node
Splits: 164 total, 164 done (100,00%)
0:12 [8,51M rows, 215MB] [687K rows/s, 17,3MB/s]

- InnerJoin[("nationkey" = "nationkey_14")][$hashvalue, $hashvalue_48] => []
    - LocalExchange[ROUND_ROBIN] () => nationkey:bigint, $hashvalue:bigint
        - RemoteExchange[REPARTITION][$hashvalue_22] => nationkey:bigint, $hashvalue_22:bigint
            - Project => [nationkey:bigint, $hashvalue_47:bigint]
                    $hashvalue_47 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("nationkey"), 0))
                - InnerJoin[("orderkey" = "orderkey_11")][$hashvalue_23, $hashvalue_44] => [nationkey:bigint]
                    - LocalExchange[ROUND_ROBIN] () => nationkey:bigint, orderkey:bigint, $hashvalue_23:bigint
                        - RemoteExchange[REPARTITION][$hashvalue_24] => nationkey:bigint, orderkey:bigint, $hashvalue_24:bigint
                            - Project => [nationkey:bigint, orderkey:bigint, $hashvalue_43:bigint]
                                    $hashvalue_43 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("orderkey"), 0))
                                - InnerJoin[("partkey" = "partkey_7") AND ("suppkey" = "suppkey_8")][$hashvalue_37, $hashvalue_38] => [nationkey:bigint, orderkey:bigint]
                                    - Project => [partkey:bigint, nationkey:bigint, suppkey:bigint, orderkey:bigint, $hashvalue_37:bigint]
                                            $hashvalue_37 := "combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("partkey"), 0)), COALESCE("$operator$hash_code"("suppkey"), 0))
                                        - InnerJoin[("suppkey" = "suppkey_3")][$hashvalue_25, $hashvalue_34] => [partkey:bigint, orderkey:bigint, suppkey:bigint, nationkey:bigint]
                                            - LocalExchange[ROUND_ROBIN] () => partkey:bigint, orderkey:bigint, suppkey:bigint, $hashvalue_25:bigint
                                                - RemoteExchange[REPARTITION][$hashvalue_26] => partkey:bigint, orderkey:bigint, suppkey:bigint, $hashvalue_26:bigint
                                                    - Project => [partkey:bigint, orderkey:bigint, suppkey:bigint, $hashvalue_33:bigint]
                                                            $hashvalue_33 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("suppkey"), 0))
                                                        - InnerJoin[("partke
```

sopel39 commented:
Even better query example:
```
create table lineitem_100000 as select * from lineitem limit 100000;

create table lineitem_50000 as select * from lineitem limit 50000;

select count(*) from (select l1.orderkey, count(*) from lineitem_100000 l1, lineitem_50000 l2 where l1.shipmode = l2.shipmode group by l1.orderkey);
```
A lot of intermediate rows are produced which are then aggregated as count. Join condition is based on VARCHARS. Customers run queries similar to that.

Before (after some warmup):
```
presto:default> select count(*) from (select l1.orderkey, count(*) from lineitem_100000 l1, lineitem_50000 l2 where l1.shipmode = l2.shipmode group by l1.orderkey);
 _col0 
-------
 24996 
(1 row)

Query 20161220_092601_00005_e2f5n, FINISHED, 1 node
Splits: 54 total, 54 done (100,00%)
0:17 [150K rows, 3,8MB] [8,84K rows/s, 229KB/s]
```

After (after some warmup):
```
presto:default> select count(*) from (select l1.orderkey, count(*) from lineitem_100000 l1, lineitem_50000 l2 where l1.shipmode = l2.shipmode group by l1.orderkey);
 _col0 
-------
 24996 
(1 row)

Query 20161220_092917_00001_yxhn5, FINISHED, 1 node
Splits: 54 total, 54 done (100,00%)
1:15 [150K rows, 3,8MB] [1,99K rows/s, 51,6KB/s]
```

martint commented on presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java:
```diff
@@ -350,6 +346,12 @@ public PlanMatchPattern withAlias(Optional<String> alias, RvalueMatcher matcher)
         return this;
     }
 
+    public PlanMatchPattern withNumberOfOutputColumns(int numberOfSymbols)
```
Avoid moving the method at the same time you make changes to the logic inside it. It makes it harder to see what's different.

sopel39 replied:
sorry, `Add support for exact symbol matching in plan matcher` commit should be squashed

martint commented on presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/SymbolCardinalityMatcher.java:
```diff
@@ -30,9 +30,15 @@
 
     @Override
-    public boolean matches(PlanNode node, Session session, Metadata metadata, ExpressionAliases expressionAliases)
+    public boolean shapeMatches(PlanNode node)
```
These changes should go in the commit that introduced this class: 37e25a974df9e559130a372b8065f047adedc13a

sopel39 replied:
squashed

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java:
```diff
@@ -36,6 +37,13 @@
     private final PlanNode right;
     private final List<EquiJoinClause> criteria;
+    /**
+     * List of output symbols produced by join. Output symbols
```
Add a check in the constructor to ensure this condition is met.

sopel39 replied:
Isn't this responsibility of `ValidateDependenciesChecker`?
For instance `ProjectNode` doesn't check the condition in constructor.

I will add `checkState` as separate commit

sopel39 commented:
rebased & applied comments

sopel39 commented:
rebased

sopel39 commented:
@martint do you have some other comments?

martint commented:
> @martint do you have some other comments?

I'm still reviewing.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java:
```diff
@@ -176,7 +187,12 @@ public PlanNode getRight()
+        return new JoinNode(getId(), type, newChildren.get(0), newChildren.get(1), criteria, outputSymbols, filter, leftHashSymbol, rightHashSymbol);
+    }
+
+    public boolean isNestedLoopJoin()
```
This should be `isCrossJoin`. Nested loops is an implementation strategy, which, technically, could be used for any type of join.

martint commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:
```diff
@@ -219,6 +219,21 @@ public void testSubqueryPruning()
+                                        tableScan("nation", ImmutableMap.of("REGIONKEY_LEFT", "regionkey", "NATIONKEY", "nationkey"))),
+                                anyTree(
+                                        tableScan("region", ImmutableMap.of("REGIONKEY_RIGHT", "regionkey"))))
+                ).withNumberOfOutputColumns(1)
```
Format as:

```java
anyTree(
    ...
)
.withNumberOfOutputColumns(1)
.withOutputs(...)
```

martint commented on presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestEliminateCrossJoins.java:
```diff
@@ -203,6 +203,10 @@ private JoinNode join(PlanNode left, PlanNode right, String... symbols)
                 left,
                 right,
                 criteria.build(),
+                ImmutableList.<Symbol>builder()
```
This seems to belong in some other commit.

sopel39 replied:
Yes, is should be part of `Add outputSymbols support to JoinNode and optimizers`

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java:
```diff
@@ -199,7 +199,23 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Set<Symbol>> context)
+            List<Symbol> outputSymbols;
+            if (node.isNestedLoopJoin()) {
+                // do not prune nested joins output since it is not supported
+                outputSymbols = ImmutableList.<Symbol>builder()
```
I haven't added output columns selection to nested loop join. I think nested loop join requires distributed mode in the first place.

I will add TODO to remove this IF when nested loop join supports output columns selection

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -403,7 +403,12 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
                 rightSource = new ProjectNode(idAllocator.getNextId(), rightSource, rightProjections.build());
 
-                output = new JoinNode(node.getId(), node.getType(), leftSource, rightSource, joinConditionBuilder.build(), newJoinFilter, node.getLeftHashSymbol(), node.getRightHashSymbol());
+                List<Symbol> outputSymbols = ImmutableList.<Symbol>builder()
```
It might produce more outputs indeed, but `PruneUnreferencedOutputs` should remove those new unused symbols. I wanted to separate concerns in this case to make make predicate pushdown cleaner. 

Anyway, I have applied your comment in additional commit: https://github.com/prestodb/presto/pull/6501/commits/7a2aaaf130ae0e238280bf51fdc4554972f5dc8e since it is easy to apply

Anyway,

sopel39 replied:
Can new join be cross join? If so, it should have all symbols from left and right sources. Therefore I think `PruneUnreferencedOutputs` should do pruning (it checks if join is cross join)

sopel39 commented:
@martint applied comments & rebased

martint commented:
Merged, thanks!
